app/src/main/java/unipd/se18/ocrcamera/PhotoTester.java

```diff
@@ -0,0 +1,449 @@
+     * @param dirName name of the directory
+     * @return the file relative to the environment and the dirName
+     * @author Pietro Prandini (g2)
+     */
+    private File getStorageDir(File environment, String dirName) {
+        // Get the directory for the user's public pictures directory.
+        File file = new File(environment, dirName);
+        if(!file.isDirectory()) {
+            Log.e(TAG, file.getAbsolutePath() + "It's not a directory");
+        } else {
+            Log.v(TAG, "Directory => " + file.getAbsolutePath());
+        }
+        return file;
+    }
+
+
+    /**
+     * Elaborate tests using threads, stores the json report to string in testReport.txt
+     * @return String in JSON format with the test's report, each object is named with the filename and contains:
+     * ingrediens, tags, notes, original photo name, confidence
+     * @author Luca Moroldo (g3)
+     */
+    public String testAndReport() {
+
+        Log.i(TAG,"testAndReport started");
+        long started = java.lang.System.currentTimeMillis();
+
+        final JSONObject jsonReport = new JSONObject();
+
+        int totalTestElements = testElements.size();
+
+        //stores the total number of tests
+        CountDownLatch countDownLatch = new CountDownLatch(totalTestElements);
+
+        int max_concurrent_tasks = Runtime.getRuntime().availableProcessors();
+        Log.i(TAG, "max_concurrent_tasks == " + max_concurrent_tasks + " (number of the available cores)");
+        Semaphore availableThread = new Semaphore(max_concurrent_tasks);
+
+        for(TestElement test : testElements){
+            try {
+                //wait for available thread
+                availableThread.acquire();
+
+                //launch thread
+                new Thread(new RunnableTest(jsonReport, test, countDownLatch, availableThread)).start();
+
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+
+        }
+
+        //wait for all tests to complete
+        try {
+            countDownLatch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+
+        long ended = java.lang.System.currentTimeMillis();
+        Log.i(TAG,"testAndReport ended (" + totalTestElements + " pics tested in " + (ended - started) + " ms)");
+
+        String report = jsonReport.toString();
+
+        //write report to file
+        try {
+            writeReportToExternalStorage(report, dirPath, "report.txt");
+        } catch (IOException e) {
+            Log.e(TAG, "Error writing report to file.");
+            e.printStackTrace();
+        }
+
+        return report;
+    }
+
+    public TestElement[] getTestElements() {
+        return testElements.toArray(new TestElement[0]);
+    }
+
+
+    /**
+     * Compare the list of ingredients extracted by OCR and the correct list of ingredients
+     * @param correct correct list of ingredients loaded from file
```

Comment:
I would have distinguished in a clearer way the "correct" parameter from the sentence that explains it otherwise it seems a repetition

Reply:
Corrected in commit 4ca6f2f